File: api/src/routes/files.ts
import { Router, Request, Response } from 'express';
import { readdir, readFile, stat } from 'fs/promises';
import { join, resolve, normalize } from 'path';
import { homedir } from 'os';

const router = Router();

function getClaudeDir(): string {
  return join(homedir(), '.claude');
}

// GET /files - Browse ~/.claude/ directory
router.get('/', async (req: Request, res: Response) => {
  try {
    const claudeDir = getClaudeDir();
    const relativePath = (req.query.path as string) || '';

    // Security: normalize and validate path is within ~/.claude/
    const requestedPath = normalize(join(claudeDir, relativePath));
    const normalizedClaudeDir = normalize(claudeDir);

    if (!requestedPath.startsWith(normalizedClaudeDir)) {
      return res.status(400).json({
        code: 'INVALID_PATH',
        message: 'Path must be within ~/.claude/'
      });
    }

    try {
      const fileStat = await stat(requestedPath);

      if (fileStat.isDirectory()) {
        // List directory contents
        const entries = await readdir(requestedPath, { withFileTypes: true });
        const items = entries.map(e => ({
          name: e.name,
          isDirectory: e.isDirectory()
        }));

        res.json({
          type: 'directory',
          path: requestedPath,
          entries: items
        });
      } else if (fileStat.isFile()) {
        // Return file content (limit to 100KB)
        if (fileStat.size > 100 * 1024) {
          return res.json({
            type: 'file',
            path: requestedPath,
            error: 'File too large (max 100KB)'
          });
        }

        const content = await readFile(requestedPath, 'utf-8');
        res.json({
          type: 'file',
          path: requestedPath,
          content
        });
      } else {
        res.status(400).json({
          code: 'INVALID_TYPE',
          message: 'Path is not a file or directory'
        });
      }
    } catch {
      res.status(404).json({
        code: 'NOT_FOUND',
        message: 'Path not found'
      });
    }
  } catch (error) {
    console.error('Error browsing files:', error);
    res.status(500).json({
      code: 'INTERNAL_ERROR',
      message: 'Failed to browse files'
    });
  }
});

export { router as filesRouter };

File: api/src/routes/files.py
"""Files routes for Claude Explorer API.

Provides generic file system browsing within the Claude data directory.
Useful for exploring raw data files not covered by specific endpoints.
"""

from pathlib import Path

from fastapi import APIRouter, HTTPException, Query

from ..models import Error, FileContent
from ..utils import get_claude_dir

router = APIRouter(prefix="/files", tags=["files"])


@router.get(
    "/",
    response_model=FileContent,
    responses={
        400: {"model": Error, "description": "Invalid path or path traversal attempt"},
        404: {"model": Error, "description": "Path not found"},
    },
)
async def browse_files(
    path: str = Query(
        "",
        description="Relative path within Claude data directory (e.g., 'projects', 'plans/my-plan.md'). Empty for root."
    )
) -> FileContent:
    """Browse the Claude data directory.

    Provides generic file system access for exploring Claude Code's data
    directory. For specific data types, prefer dedicated endpoints
    (projects, sessions, plans, skills, etc.).

    Security:
    - Path traversal is prevented (cannot access files outside data directory)
    - File content limited to 100KB

    For directories:
        Returns list of entries with name and isDirectory flag

    For files:
        Returns file content as string (max 100KB)
        Returns error message if file exceeds size limit

    Args:
        path: Relative path within the data directory

    Returns:
        FileContent with type ('file' or 'directory'), path, and either
        content (for files) or entries (for directories)

    Raises:
        400: Path traversal attempt or invalid path type
        404: Path not found
    """
    claude_dir = get_claude_dir()

    # Security: normalize and validate path is within ~/.claude/
    requested_path = (claude_dir / path).resolve()

    if not str(requested_path).startswith(str(claude_dir.resolve())):
        raise HTTPException(status_code=400, detail="Path must be within Claude data directory")

    if not requested_path.exists():
        raise HTTPException(status_code=404, detail="Path not found")

    if requested_path.is_dir():
        # List directory contents
        entries = [
            {"name": e.name, "isDirectory": e.is_dir()}
            for e in requested_path.iterdir()
        ]
        return {
            "type": "directory",
            "path": str(requested_path),
            "entries": entries,
        }
    elif requested_path.is_file():
        # Return file content (limit to 100KB)
        if requested_path.stat().st_size > 100 * 1024:
            return {
                "type": "file",
                "path": str(requested_path),
                "error": "File too large (max 100KB)",
            }

        content = requested_path.read_text()
        return {
            "type": "file",
            "path": str(requested_path),
            "content": content,
        }
    else:
        raise HTTPException(status_code=400, detail="Path is not a file or directory")
